File: src/LorrisAnalyzer/DataWidgets/ScriptWidget/examples/snake.py
# Python implementation of Snake game for Lorris
# Controls: click to terminal and press:
#    r - start/stop
#    p - pause
#    wasd - control snake's movement

from random import *

MOVE_UP    = 0
MOVE_DOWN  = 1
MOVE_LEFT  = 2
MOVE_RIGHT = 3

STEP_LEN = script.height

terminal.clear()
lorris.resizeWidget(script, 90, 90);

timer = lorris.newTimer()
direction = MOVE_RIGHT
run = False
cookie = None
cookieColor = []
tail = []

scoreW = lorris.newWidget(WIDGET_NUMBER, 'Score', 150, 0)
lorris.moveWidget(scoreW, area.width - 150, 0)
score = 0

moveQueue = []

def startStop():
    global run
    if run:
        timer.stop();
    else:
        start();
    run = not run

def pause():
    if not run:
        return
    if timer.active:
        timer.stop()
    else:
        timer.start(400)

def onKeyPress(key):
    if key == 'r': startStop()
    elif key == 'p': pause()

    if not timer.active:
        return

    hasTail = (tail != None and len(tail) != 0)

    if key == 'w' and (not hasTail or direction != MOVE_DOWN):
        moveQueue.append(MOVE_UP)
    elif key == 'a' and (not hasTail or direction != MOVE_RIGHT):
        moveQueue.append(MOVE_LEFT)
    elif key == 's' and (not hasTail or direction != MOVE_UP):
        moveQueue.append(MOVE_DOWN)
    elif key == 'd' and (not hasTail or direction != MOVE_LEFT):
        moveQueue.append(MOVE_RIGHT)

def start():
    global direction
    global scoreW
    global tail
    global moveQueue
    global cookie
    global timer

    lorris.moveWidget(script, 0, 0);
    
    direction = MOVE_RIGHT
    score = 0
    scoreW.setValue(0)

    while len(tail) != 0:
        tail.pop().remove()

    del moveQueue[:]

    if cookie != None:
        cookie.remove();

    spawnCookie();

    timer.start(400);

def moveAll():
    global moveQueue
    global direction
    if len(moveQueue) == 0:
        moveQueue.append(direction);
    
    while len(moveQueue) != 0:
        direction = moveQueue.pop(0)
        prevPos = [script.x + 5, script.y + 5 ]
        move(script, direction)

        for i in tail:
            pos = [i.x, i.y]
            lorris.moveWidget(i, prevPos[0], prevPos[1])
            prevPos = pos

    if not checkCollision():
        return

    if rectOverlaps(script, cookie):
        consumeCookie()

timer.connect('timeout()', moveAll);

def move(widget, dir):
    x = widget.x
    y = widget.y

    if dir == MOVE_UP: y -= STEP_LEN
    elif dir == MOVE_DOWN: y += STEP_LEN
    elif dir == MOVE_LEFT: x -= STEP_LEN
    elif dir == MOVE_RIGHT: x += STEP_LEN

    if y < 0: y = area.height + y
    elif y > area.height: y = y - area.height

    if x < 0: x = area.width + x
    elif x > area.width: x = x - area.width

    lorris.moveWidget(widget, x, y)

def spawnCookie():
    global cookieColor
    global cookie

    x = randint(0, area.width-110)
    y = randint(0, area.height-110)

    cookie = lorris.newWidget(WIDGET_COLOR, "Cookie", 110, 110)
    cookieColor = [ randint(0, 255), randint(0, 255), randint(0, 255) ]
    cookie.setValue(cookieColor[0], cookieColor[1], cookieColor[2])

    lorris.moveWidget(cookie, x, y)

def rectOverlaps(first, second):
    f_xw = first.x + first.width
    f_yh = first.y + first.height
    s_xw = second.x + second.width;
    s_yh = second.y + second.height;

    xOverlap = valueInRange(first.x, second.x, s_xw) or valueInRange(second.x, first.x, f_xw)
    yOverlap = valueInRange(first.y, second.y, s_yh) or valueInRange(second.y, first.y, f_yh)
    return xOverlap and yOverlap

def valueInRange(val, min, max):
    return val >= min and val <= max

def consumeCookie():
    global tail
    global score
    global tail
    global scoreW

    tailpart = lorris.newWidget(WIDGET_COLOR, "T", script.width-10, script.height-10)
    tailpart.setValue(cookieColor[0], cookieColor[1], cookieColor[2])
    
    cookie.remove()
    spawnCookie()

    p = [0, 0]
    if len(tail) != 0:
        p[0] = tail[len(tail)-1].x
        p[1] = tail[len(tail)-1].y
    else:
        p[0] = script.x + 5
        p[1] = script.y + 5

    lorris.moveWidget(tailpart, p[0], p[1])
    tail.append(tailpart)

    score += 10
    scoreW.setValue(score)

def checkCollision():
    global run
    global timer
    for i in tail:
        if rectOverlaps(script, i):
            run = False
            timer.stop()
            lorris.throwException("Game Over!")
            return False
    return True

File: src/LorrisAnalyzer/DataWidgets/ScriptWidget/examples/snake.js
// JavaScript implementation of Snake game for Lorris
// Controls: click to terminal and press:
//    r - start/stop
//    p - pause
//    wasd - control snake's movement

var MOVE_UP    = 0;
var MOVE_DOWN  = 1;
var MOVE_LEFT  = 2;
var MOVE_RIGHT = 3;

var STEP_LEN = script.height;

var timer = newTimer();
var direction = MOVE_RIGHT;
var run = false;
var tail;
var cookie = null;

var scoreW = newNumberWidget("Score", 150, 0);
moveWidget(scoreW, area.width - 150, 0);
var score = 0;

var moveQueue;

resizeWidget(script, 90, 90);
clearTerm();

function Pos(x, y)
{
    this.x = x;
    this.y = y;
}

function onKeyPress(key) {
    switch(key)
    {
        case "r":
            if(run)
                timer.stop();
            else
                start();
            run = !run;
            return;
        case "p":
        {
            if(!run)
                return;
            
            if(timer.active)
                timer.stop();
            else
                timer.start(400);
            return;
        }
    }

    if(!timer.active)
        return;

    var hasTail = tail != null && tail.lenght != 0;

    switch(key)
    {
        case "w":
            if(!hasTail || direction != MOVE_DOWN)
                moveQueue.push(MOVE_UP);
            break;
        case "a":
            if(!hasTail || direction != MOVE_RIGHT)
                moveQueue.push(MOVE_LEFT);
            break;
        case "s":
            if(!hasTail || direction != MOVE_UP)
                moveQueue.push(MOVE_DOWN);
            break;
        case "d":
            if(!hasTail || direction != MOVE_LEFT)
                moveQueue.push(MOVE_RIGHT);
            break;
    }
}

function start()
{
    moveWidget(script, 0, 0);

    direction = MOVE_RIGHT;
    score = 0;
    scoreW.setValue(0);
    
    for(var i = 0; tail != null && i < tail.length; ++i)
        tail[i].remove();
    tail = new Array();

    moveQueue = new Array();

    if(cookie != null)
        cookie.remove();

    spawnCookie();
    
    timer.start(300);
}

function moveAll()
{
    if(moveQueue.length == 0)
        moveQueue.push(direction);

    while(moveQueue.length != 0)
    {
        direction = moveQueue.shift();
        var prevPos = new Pos(script.x + 5, script.y + 5);
        move(script, direction);
    
        for(var i = 0; i < tail.length; ++i)
        {
            var pos = new Pos(tail[i].x, tail[i].y);
            moveWidget(tail[i], prevPos.x, prevPos.y);
            prevPos = pos;
        }
    }

    if(!checkCollision())
        return;

    if(rectOverlaps(script, cookie))
        consumeCookie();
}
timer.timeout.connect(moveAll);

function move(widget, dir)
{
    var x = widget.x;
    var y = widget.y;
    switch(dir)
    {
        case MOVE_UP:
            y -= STEP_LEN;
            break;
        case MOVE_DOWN:
            y += STEP_LEN;
            break;
        case MOVE_LEFT:
            x -= STEP_LEN;
            break;
        case MOVE_RIGHT:
            x += STEP_LEN;
            break;
    }

    if(y < 0)                y = area.height + y;
    else if(y > area.height) y = y - area.height;

    if(x < 0)               x = area.width + x;
    else if(x > area.width)    x = x - area.width;

    moveWidget(widget, x, y);
}

var cookieColor;
function spawnCookie()
{
    var x = rand(area.width-110);
    var y = rand(area.height-110);

    cookie = newColorWidget("Cookie", 110, 110);
    cookieColor = new Array(rand(255), rand(255), rand(255));
    cookie.setValue(cookieColor[0], cookieColor[1], cookieColor[2]);

    moveWidget(cookie, x, y);
}

function rectOverlaps(first, second)
{
    var f_xw = first.x + first.width;
    var f_yh = first.y + first.height;
    var s_xw = second.x + second.width;
    var s_yh = second.y + second.height;
    
    var xOverlap = valueInRange(first.x, second.x, s_xw) || valueInRange(second.x, first.x, f_xw);
    var yOverlap = valueInRange(first.y, second.y, s_yh) || valueInRange(second.y, first.y, f_yh);

    return (xOverlap && yOverlap);
}

function valueInRange(val, min, max)
{
    return (val >= min) && (val <= max);
}

function consumeCookie()
{
    var tailpart = newColorWidget("T", script.width-10, script.height-10);
    tailpart.setValue(cookieColor[0], cookieColor[1], cookieColor[2]);

    cookie.remove();
    spawnCookie();

    var p = new Pos(0, 0);
    if(tail.length != 0)
    {
        p.x = tail[tail.length - 1].x;
        p.y = tail[tail.length - 1].y;
    }
    else
    {
        p.x = script.x + 5;
        p.y = script.y + 5;
    }

    moveWidget(tailpart, p.x, p.y);
    tail.push(tailpart);

    score += 10;

    scoreW.setValue(score);
}

function checkCollision()
{
    for(var i = 0; i < tail.length; ++i)
    {
        if(rectOverlaps(script, tail[i]))
        {
            run = false;
            timer.stop();
            throwException("Game Over!");
            return false;
        }
    }
    return true;
}

function rand(max)
{
    return Math.floor(Math.random() * max);
}
